---
title: "Beta_diversity_rumen"
author: "Chiara Gini"
date: "20/12/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("knitr")
library("dplyr")
library("tidyr")
library("broom")
library("vegan")
library("ggplot2")
library("reshape2")
library("rmarkdown")
library("tidyverse")
library("ggfortify")
library("data.table")
```

### Beta diversity analysis of samples of ruminal liquid of 4 animals (collected at the sloather house): I1, I2, I3 and I4. Control group consists in ruminal liquid + diet, treatments consist of control and AE1, AE sintethyc 1, Carvacrol, p-cymene, gamma-terpinene. 
 
# Comparison 1) Beta diversity: RUMINAL LIQUID VS CONTROL (RUMINAL LIQUID +DIET)

Between-sample variability from 16S rRna-gene sequencing data.

```{r, echo=FALSE}
metadata <- fread("/home/mycelium/milkqua_project/Rumen/mapping_file.csv")
names(metadata)[1] <- "sample"
names(metadata)[3] <- "timepoint"
metadata$treatment[metadata$treatment=="0.0"] <- "Control"
metadata$treatment[metadata$treatment==""] <- "Ruminal liquid"
```

## Clustering

### PCA

```{r, echo=FALSE, message=FALSE}
matrice= read.table("/home/mycelium/milkqua_project/Rumen/results/beta_diversity/bray_curtis_CSS_normalized_otu_table.txt", row.names=1, header=T)
names(matrice) <- gsub("X","",names(matrice))

vec <- filter(metadata, timepoint == "MILQUA") %>% select(sample) %>% pull()
vex <- names(matrice) %in% vec
mat_rumen = matrice[vex,vex]

mat_rumen$treatment <- as.character(metadata$treatment[match(row.names(mat_rumen),metadata$sample)])
mat_rumen$timepoint <- as.character(metadata$timepoint[match(row.names(mat_rumen),metadata$sample)])

matx= data.matrix(select(mat_rumen, -c(treatment,timepoint)))

## MDS
mds <- cmdscale(as.dist(matx))
```

Clustering by treatment (PCA) - ruminal liquid & control:

```{r, echo=FALSE, message=FALSE, warning=FALSE}
mat_rumen <- filter(mat_rumen, treatment != "AE1", treatment !="AE sintético 1",treatment != "Carvacrol", treatment !="p-cymene", treatment !="γ-terpinene")
matx= data.matrix(select(mat_rumen, -c(treatment)))
p1 <- autoplot(prcomp(1-matx), data=mat_rumen, colour = "treatment", label = TRUE, label.size = 3, frame = TRUE, frame.type = 'norm')
#p1
ggsave (p1, height = 4, width = 8, filename = "PCA_clustering_by_treatment_CTRLvsRuminalFliud_rumen.png")
```
### MDS

#### Clustering by treatment (MDS): **NMDS1 vs NMDS2**

```{r, echo=FALSE, results='hide'}
matx0= data.matrix(select(mat_rumen, -c(treatment,timepoint)))
ruminal.mds= metaMDS(matx0, k=3) #function metaMDS in Vegan

hull_f <- function(df) {

  temp <- data.frame(NULL)
  for (ll in unique(df$group)) {

    nn <- df[df$group == ll,][chull(df[df$group == ll, c("NMDS1","NMDS2","NMDS3")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

ruminal.scores <- as.data.frame(scores(ruminal.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
ruminal.scores$group <- mat_rumen$treatment #  add the grp variable created earlier

hull.data <- hull_f(ruminal.scores)

g <- ggplot(data=ruminal.scores, aes(x=NMDS1,y=NMDS2))
g <- g + coord_equal()
g = g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS2,fill=group,group=group),alpha=0.30)
g = g + geom_point(data=ruminal.scores,aes(x=NMDS1,y=NMDS2,shape=group,colour=group),size=4)
g = g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g <- g + ggtitle("Treatment")
#g
print(g)
```

#### Clustering by type (MDS): **NMDS1 vs NMDS3**

```{r, echo=FALSE, results='hide'}
g <- ggplot(data=ruminal.scores, aes(x=NMDS1,y=NMDS3))
g <- g + coord_equal()
g = g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS3,fill=group,group=group),alpha=0.30)
g = g + geom_point(data=ruminal.scores,aes(x=NMDS1,y=NMDS3,shape=group,colour=group),size=4)
g = g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g <- g + ggtitle("Source ")
print(g)
```

#### Clustering by type (MDS): **NMDS2 vs NMDS3**

```{r, echo=FALSE, results='hide'}
g <- ggplot(data=ruminal.scores, aes(x=NMDS2,y=NMDS3))
g <- g + coord_equal()
g = g + geom_polygon(data=hull.data,aes(x=NMDS2,y=NMDS3,fill=group,group=group),alpha=0.30)
g = g + geom_point(data=ruminal.scores,aes(x=NMDS2,y=NMDS3,shape=group,colour=group),size=4)
g = g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g <- g + ggtitle("Source")
print(g)
```


### Clustering by treatment by timepoint (MDS):

```{r, echo=FALSE, results='hide'}
## HULL - timepoint + treatment
ruminal.mds= metaMDS(matx) #function metaMDS in Vegan

ruminal.scores <- as.data.frame(scores(ruminal.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
ruminal.scores$timepoint <- mat_rumen$timepoint #  add the grp variable created earlier
ruminal.scores$treatment <- mat_rumen$treatment


hull_f <- function(df) {

  temp <- data.frame(NULL)
  for (ll in unique(df$treatment)) {

    nn <- df[df$type == ll,][chull(df[df$type == ll, c("NMDS1","NMDS2")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

hull.data <- hull_f(ruminal.scores)

g= ggplot(data=ruminal.scores, aes(x=NMDS1,y=NMDS2))
# g <- g + geom_point(aes(shape=group,colour=group),size=3)
# g= g + coord_equal()
g= g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS2,fill=treatment,group=treatment),alpha=0.30)
g= g + geom_point(data=ruminal.scores,aes(x=NMDS1,y=NMDS2,shape=treatment,colour=treatment),size=4)
g= g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g= g + facet_wrap(~timepoint, scales = "free")
# g <- g + theme(legend.position="none")
g= g + ggtitle("TREATMENT x TIMEPOINT")
print(g)
ggsave(g, height = 6, width = 6, filename = "beta_diversity_treatmentxtimepoint_rumen.png")
```

## Significance of between-group distances

Significance values based on permuted analysis of variance (999 permutations), repeated 100 times.

```{r,  echo=FALSE}
pv_treatment <- replicate(100,
  adonis2(matx ~ mat_rumen$treatment, permutations = 999)$"Pr(>F)"[1],
  simplify = "vector"
)

permanova <- function(mat,term,perms) {

  perm <- how(nperm = sample(perms,1))
  obj <- adonis(mat ~ term, add = TRUE, parallel = TRUE, permutations = perm)
  return(obj$aov.tab$`Pr(>F)`[1])
}
```

- between treatments: p-value = **`r mean(pv_treatment)`** (std = `r sd(pv_treatment)`)

## Below the ANOVA table from the model:

$$
X = \mu + treatment + timepoint + treatment*type + e
$$

```{r, echo=FALSE, message=FALSE, warning=FALSE}
source("/home/mycelium/milkqua_project/Rumen/pairwise_adonis.r")

obj <- adonis(matx ~ mat_rumen$treatment)
row.names(obj$aov.tab) <- c("treatment","residuals","total")
kable(obj$aov.tab)
```

### Going 3D

```{r, echo=FALSE, warning=FALSE, message=FALSE}
## HULL - timepoint + treatment
ruminal.mds= metaMDS(matx, k = 3) #function metaMDS in Vegan

ruminal.scores <- as.data.frame(scores(ruminal.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
ruminal.scores$timepoint <- mat_rumen$timepoint #  add the grp variable created earlier
ruminal.scores$treatment <- mat_rumen$treatment


hull_f <- function(df) {
  
  temp <- data.frame(NULL)
  for (ll in unique(df$treatment)) {
    
    nn <- df[df$timepoint == ll,][chull(df[df$timepoint == ll, c("NMDS1","NMDS2","NMDS3")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

hull.data <- hull_f(ruminal.scores)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
library("plotly")

p <- plot_ly(data = ruminal.scores, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p

if(pandoc_available()) {
  
  htmlwidgets::saveWidget(as_widget(p), "beta_timepoint_3d.html")
}
```

# Comparison 2) Beta diversity: CONTROL VS TREATMENTS (control and treatments have all the same base of ruminal liquid+diet)

Between-sample variability from 16S rRna-gene sequencing data.

```{r, echo=FALSE}
metadata <- fread("/home/mycelium/milkqua_project/Rumen/mapping_file.csv")
names(metadata)[1] <- "sample"
names(metadata)[3] <- "timepoint"
metadata$treatment[metadata$treatment=="0.0"] <- "Control"
```

## Clustering

### PCA

```{r, echo=FALSE, message=FALSE}
matrice= read.table("/home/mycelium/milkqua_project/Rumen/results/beta_diversity/bray_curtis_CSS_normalized_otu_table.txt", row.names=1, header=T)
names(matrice) <- gsub("X","",names(matrice))

vec <- filter(metadata, timepoint == "MILQUA") %>% select(sample) %>% pull()
vex <- names(matrice) %in% vec
mat_rumen = matrice[vex,vex]

mat_rumen$treatment <- as.character(metadata$treatment[match(row.names(mat_rumen),metadata$sample)])
mat_rumen$timepoint <- as.character(metadata$timepoint[match(row.names(mat_rumen),metadata$sample)])

matx= data.matrix(select(mat_rumen, -c(treatment,timepoint)))

## MDS
mds <- cmdscale(as.dist(matx))
```

Clustering by timepoint (PCA), removing ruminal liquid samples (I1 to I4):

```{r, echo=FALSE, message=FALSE, warning=FALSE}
mat_rumen <- filter(mat_rumen, treatment !="")
matx=data.matrix(select(mat_rumen, -c(treatment,timepoint)))
p1 <- autoplot(prcomp(1-matx), data=mat_rumen, colour = "timepoint", label = TRUE, label.size = 3, frame = TRUE, frame.type = 'norm')
p1
ggsave (p1, height = 4, width = 8, filename = "clustering_by_timepoints_PCA_rumen_treatments.png")
```

Clustering by treatment (PCA)

```{r, echo=FALSE}
matx0= data.matrix(select(mat_rumen, -c(treatment)))
p <- autoplot(prcomp(1-matx0), data=mat_rumen, colour = "treatment", label = TRUE, label.size = 3, frame = TRUE, frame.type='norm')
p

ggsave (p, height = 4, width = 8, filename = "clustering_by_treatment_PCA_rumen_treatments.png")
```

### MDS

#### Clustering by treatment (MDS): **NMDS1 vs NMDS2**

```{r, echo=FALSE, results='hide'}
ruminal.mds= metaMDS(matx0, k=3) #function metaMDS in Vegan

hull_f <- function(df) {

  temp <- data.frame(NULL)
  for (ll in unique(df$group)) {

    nn <- df[df$group == ll,][chull(df[df$group == ll, c("NMDS1","NMDS2","NMDS3")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

ruminal.scores <- as.data.frame(scores(ruminal.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
ruminal.scores$group <- mat_rumen$treatment #  add the grp variable created earlier

hull.data <- hull_f(ruminal.scores)

g <- ggplot(data=ruminal.scores, aes(x=NMDS1,y=NMDS))
g <- g + coord_equal()
g = g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS2,fill=group,group=group),alpha=0.30)
g = g + geom_point(data=ruminal.scores,aes(x=NMDS1,y=NMDS2,shape=group,colour=group),size=4)
g = g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g <- g + ggtitle("Treatment")
print(g)
```

Clustering by type (MDS): **NMDS1 vs NMDS3**

```{r, echo=FALSE, results='hide'}
g <- ggplot(data=ruminal.scores, aes(x=NMDS1,y=NMDS3))
g <- g + coord_equal()
g = g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS3,fill=group,group=group),alpha=0.30)
g = g + geom_point(data=ruminal.scores,aes(x=NMDS1,y=NMDS3,shape=group,colour=group),size=4)
g = g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g <- g + ggtitle("Source ")
print(g)
```

Clustering by type (MDS): **NMDS2 vs NMDS3**

```{r, echo=FALSE, results='hide'}
g <- ggplot(data=ruminal.scores, aes(x=NMDS2,y=NMDS3))
g <- g + coord_equal()
g = g + geom_polygon(data=hull.data,aes(x=NMDS2,y=NMDS3,fill=group,group=group),alpha=0.30)
g = g + geom_point(data=ruminal.scores,aes(x=NMDS2,y=NMDS3,shape=group,colour=group),size=4)
g = g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g <- g + ggtitle("Source")
print(g)
```

### Clustering by treatment by timepoint (MDS):

```{r, echo=FALSE, results='hide'}
## HULL - timepoint + treatment
ruminal.mds= metaMDS(matx) #function metaMDS in Vegan

ruminal.scores <- as.data.frame(scores(ruminal.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
ruminal.scores$timepoint <- mat_rumen$timepoint #  add the grp variable created earlier
ruminal.scores$treatment <- mat_rumen$treatment


hull_f <- function(df) {

  temp <- data.frame(NULL)
  for (ll in unique(df$treatment)) {

    nn <- df[df$type == ll,][chull(df[df$type == ll, c("NMDS1","NMDS2")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

hull.data <- hull_f(ruminal.scores)

g= ggplot(data=ruminal.scores, aes(x=NMDS1,y=NMDS2))
# g <- g + geom_point(aes(shape=group,colour=group),size=3)
# g= g + coord_equal()
g= g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS2,fill=treatment,group=treatment),alpha=0.30)
g= g + geom_point(data=ruminal.scores,aes(x=NMDS1,y=NMDS2,shape=treatment,colour=treatment),size=4)
g= g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g= g + facet_wrap(~timepoint, scales = "free")
# g <- g + theme(legend.position="none")
g= g + ggtitle("TREATMENT x TIMEPOINT")
print(g)
#ggsave(g, height = 6, width = 6, filename = "beta_diversity_treatmentxtimepoint_rumen_treatments.png")
```

## Significance of between-group distances

Significance values based on permuted analysis of variance (999 permutations), repeated 100 times.

```{r, echo=FALSE}
pv_treatment <- replicate(100,
  adonis2(matx ~ mat_rumen$treatment, permutations = 999)$"Pr(>F)"[1],
  simplify = "vector"
)

permanova <- function(mat,term,perms) {

  perm <- how(nperm = sample(perms,1))
  obj <- adonis(mat ~ term, add = TRUE, parallel = TRUE, permutations = perm)
  return(obj$aov.tab$`Pr(>F)`[1])
}
```

Between treatments: p-value = **`r mean(pv_treatment)`** (std = `r sd(pv_treatment)`)

Below the ANOVA table from the model:

$$
X = \mu + treatment + timepoint + treatment*type + e
$$

```{r, echo=FALSE, message=FALSE, warning=FALSE}
source("/home/mycelium/milkqua_project/Rumen/pairwise_adonis.r")

obj <- adonis(matx ~ mat_rumen$treatment)
row.names(obj$aov.tab) <- c("treatment","residuals","total")
kable(obj$aov.tab)
```

## going 3D

```{r, echo=FALSE, warning=FALSE, message=FALSE}
## HULL - timepoint + treatment
ruminal.mds= metaMDS(matx, k = 3) #function metaMDS in Vegan

ruminal.scores <- as.data.frame(scores(ruminal.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
ruminal.scores$timepoint <- mat_rumen$timepoint #  add the grp variable created earlier
ruminal.scores$treatment <- mat_rumen$treatment


hull_f <- function(df) {
  
  temp <- data.frame(NULL)
  for (ll in unique(df$treatment)) {
    
    nn <- df[df$timepoint == ll,][chull(df[df$timepoint == ll, c("NMDS1","NMDS2","NMDS3")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

hull.data <- hull_f(ruminal.scores)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
## 3D plot
library("plotly")

p <- plot_ly(data = ruminal.scores, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p

if(pandoc_available()) {
  
  htmlwidgets::saveWidget(as_widget(p), "beta_timepoint_3d.html")
}
```

### Treatment AE1

```{r, echo=FALSE, warning=FALSE, message=FALSE}
temp <- filter(ruminal.scores, treatment == "AE1")

p <- plot_ly(data = temp, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```


### timepoint AE sintético 1

```{r, echo=FALSE, warning=FALSE, message=FALSE}
temp <- filter(ruminal.scores, treatment == "AE sintético 1")

p <- plot_ly(data = temp, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

### treatment Carvacrol

```{r, echo=FALSE, warning=FALSE, message=FALSE}
temp <- filter(ruminal.scores, treatment == "Carvacrol")

p <- plot_ly(data = temp, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

### treatment p-cymene

```{r, echo=FALSE, warning=FALSE, message=FALSE}
temp <- filter(ruminal.scores, treatment == "p-cymene")

p <- plot_ly(data = temp,  
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

### treatment γ-terpinene

```{r, echo=FALSE, warning=FALSE, message=FALSE}
temp <- filter(ruminal.scores, treatment == "γ-terpinene")

p <- plot_ly(data = temp,  
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

### treatment Control

```{r, echo=FALSE, warning=FALSE, message=FALSE}
temp <- filter(ruminal.scores, treatment == "Control")

p <- plot_ly(data = temp,  
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```
