---
title: "MILKQUA project: milk samples - Metabarcoding Stats"
author: "Chiara Gini"
date: "27/12/2021"
output:
  html_document: default
---

```{r setup, include=FALSE}
library("knitr")
library("broom")
# library("plyr")
library("dplyr")
library("tidyr")
library("ggplot2")
library("reshape2")
library("data.table")
```

## 16S rRNA-gene sequencing - bovine milk

### Joining paired-end reads

The first step is joining the sequenced paired-end reads.

```{r reads, echo=FALSE, warning=FALSE, message=FALSE}
joined_pairs <- read.csv("/home/oem/milkqua_microbiome/Results/results/readsPerSample",sep="\t",header = FALSE,skip = 1, stringsAsFactors = FALSE)
nReads <- as.numeric(joined_pairs$V1[seq(2,nrow(joined_pairs)-1,by=2)])
samples <- joined_pairs$V1[seq(1,nrow(joined_pairs)-1,by=2)]

joined_pairs <- data.frame("sample"=samples,"reads"=nReads)
joined_pairs$reads <- joined_pairs$reads/4 ## dividing by 4 (fastq file format)
```

Overall, the average number of input reads (both R1 and R2) was $\mathbf{`r as.integer(mean(joined_pairs$reads))`}$.

```{r, echo=FALSE}
D <- joined_pairs %>%
  filter(!grepl("UndeterminedUndetermined",sample)) %>%
  dplyr::summarize(
    n=n(),
    avgReads=mean(reads),
    sdReads = sd(reads),
    maxReads=round(max(reads),3),
    minReads=round(min(reads),3)
)

kable(D)
```

Below, the barplot of joined reads per sample:

```{r pressure, echo=FALSE}
mPaired <- reshape2::melt(joined_pairs %>% filter(!grepl("UndeterminedUndetermined",sample)),id.vars=c("sample"))
p <- ggplot(mPaired,aes(x=sample,y=value)) + geom_bar(aes(fill=sample), stat = "identity")
p <- p + guides(fill=FALSE)
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1, size=3))
print(p)
```

### Multiple Split Library

Reads were then filtered for:

- maximum bad run length: **3** (max n. of consecutive low quality -based on Phred- base calls)
- minimum read leangth fraction: **0.75**
- Phred quality threshold: **19**


```{r, echo=FALSE, warning=FALSE}
inputSeqs <- read.table("/home/oem/milkqua_microbiome/Results/results/inputSeqs",sep=":")
outputSeqs <- read.table("/home/oem/milkqua_microbiome/Results/results/outputSeqs", sep="\t")
samples <- read.table("//home/oem/milkqua_microbiome/Results/results/samples.txt", sep=":")

trim <- function (x) gsub("^\\s+|\\s+$", "", x)

samples$V2 <- gsub("/storage/core/metabarcoding_milkqua/Analysis/multiple_join_paired_ends/","",samples$V2)
samples$V1 <- NULL
samples$V3 <- NULL
samples$V2 <- gsub("/.*","",samples$V2)
samples$V2 <- trim(samples$V2)
samples$V2 <- gsub("^\\d*-","",samples$V2)
samples$V2 <- gsub("_S.*$","",samples$V2)

samples <- samples %>%
  rowwise() %>%
  mutate(n=ifelse(nchar(as.character(V2))==9,gsub("[0-9]{1}$","",V2),gsub("[0-9]{2}$","",V2)))

samples$V2 <- NULL
names(samples)[1] <- "sample"

seqs <- cbind.data.frame(samples,inputSeqs$V2,outputSeqs$V2)
names(seqs) <- c("sample","input","output")

# detach("package:plyr", unload=TRUE) 

seqs <- seqs %>%
  group_by(sample) %>%
  summarise(inpSeqs=sum(input),outSeqs=sum(output))

seqs <- seqs %>%
  mutate(loss=(inpSeqs-outSeqs)/inpSeqs) %>%
  arrange(loss)

kable(seqs)
# write.table(seqs, file="afterSplit.csv",sep=",",quote=FALSE,row.names=FALSE)
```

Below basic stats on the total number of reads in input and output (after filtering):

```{r, echo=FALSE}
D <- seqs

DD <- D %>%
  # .(group),
  summarize(
  "inputSeq"=sum(inpSeqs),
  "outputSeqs"=sum(outSeqs),
  "avgInp"=mean(inpSeqs),
  "avgOutput"=mean(outSeqs),
  "stdInp"=sd(inpSeqs),
  "stdOutput"=sd(outSeqs,na.rm=TRUE)
)

kable(DD)
```

And the average proportion of retained reads for subsequent analysis:

```{r, echo=FALSE}
d1 <- D %>%
  mutate(retained=outSeqs/inpSeqs) %>%
  summarize(maxRetained=max(retained),minRetained=min(retained),avgRetained=mean(retained))

kable(d1)

# d2 <- D %>%
#   summarize(sum(inpSeqs),sum(outSeqs))
# 
# kable(d2)
```


```{r, echo=FALSE}
mS <- reshape2::melt(seqs,id.vars = c("sample","loss"), variable.name = "seq", value.name = "reads")
mS <- mS %>%
  filter(!grepl("Undetermined",sample)) %>%
  arrange(reads)

seqs <- seqs %>%
  mutate(nloss = inpSeqs-outSeqs) %>%
  arrange(nloss)

mS$sample <- factor(mS$sample, levels = seqs$sample)

p <- ggplot(mS, aes(x=sample,y=reads,group=seq)) + geom_line(aes(colour=seq))
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1, size=4))
p
```

### Summary stats

```{r, echo=FALSE}
mapfile <- fread("/home/oem/milkqua_microbiome/Results/mapping_file_milk_alltimepoints_skin_T0beforeafter.csv", header = TRUE)
mapfile$sample <- paste("Sample",sapply(mapfile$`#SampleID`, function(x) substr(x,1,nchar(x)/2)),sep="_")
seqs <- seqs %>%
  inner_join(mapfile, by = "sample")
```

#### Per type

```{r, echo=FALSE}
seqs %>%
  summarise(avg_seq = mean(outSeqs), std_seqs = sd(outSeqs)) %>%
  kable()

seqs %>%
  group_by(sample_type) %>%
  summarise(avg_seq = mean(outSeqs), std_seqs = sd(outSeqs)) %>%
  kable()
```

```{r, echo=FALSE}
seqs %>%
  do(glance(lm(outSeqs~sample_type,.))) %>%
  kable()
```

```{r, echo=FALSE}
ggplot(seqs, aes( x = as.factor(sample_type), y = outSeqs)) + geom_boxplot(aes(fill=as.factor(sample_type)))
```

#### Per ref

```{r, echo=FALSE}
seqs %>%
    group_by(ref) %>%
  summarise(N=n(), avg_seq = mean(outSeqs), std_seqs = sd(outSeqs)) %>%
  kable()
```

```{r, echo=FALSE}
seqs %>%
  do(tidy(lm(outSeqs~ref,.))) %>%
  kable()
```

```{r, echo=FALSE}
library("rstatix")
# pwc <- seqs %>% tukey_hsd(outSeqs ~ as.factor(timepoint))
# print(pwc)
```


```{r, echo=FALSE}
ggplot(seqs, aes( x = as.factor(ref), y = outSeqs)) + geom_boxplot(aes(fill=as.factor(ref)))
```


