---
title: "Beta_diversity_SKINSWAB"
author: "Chiara Gini"
date: "27/12/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
library("knitr")
library("dplyr")
library("tidyr")
library("broom")
library("vegan")
library("ggplot2")
library("reshape2")
library("rmarkdown")
library("tidyverse")
library("ggfortify")
library("data.table")
library("ggforce")
library("plotly")
library("lattice")
```

## Beta diversity

From 16S rRNA-gene sequencing of cow SWAB microbiome samples:
- 3 timepoints (T0 BEFORE, T0 AFTER AND T8) for SWAB samples and two treatments (oil, treated/untreated)


Between-sample variability from 16S rRna-gene sequencing data.

```{r beta, echo=FALSE}
metadata <- fread("/home/oem/milkqua_microbiome/Results/mapping_file_milk_alltimepoints_skin_T0beforeafter.csv")
names(metadata)[1] <- "sample"
```

## Clustering

### PCA

```{r pressure, echo=FALSE, message=FALSE}
matrice= read.table("/home/oem/milkqua_microbiome/Results/results/beta_diversity/bray_curtis_CSS_normalized_otu_table.txt", row.names=1, header=T)
names(matrice) <- gsub("X","",names(matrice))

vec <- filter(metadata, sample_type == "swab") %>% select(sample) %>% pull()
vex <- names(matrice) %in% vec
mat_swab = matrice[vex,vex]

# 
mat_swab$treatment <- as.character(metadata$treatment[match(row.names(mat_swab),metadata$sample)])
mat_swab$timepoint <- as.character(metadata$timepoint[match(row.names(mat_swab),metadata$sample)])

matx= data.matrix(select(mat_swab, -c(treatment,timepoint)))

## MDS
mds <- cmdscale(as.dist(matx))
```

Clustering by timepoint (PCA):

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p1 <- autoplot(prcomp(1-matx), data=mat_swab, colour = "timepoint", label = TRUE, label.size = 3, frame = TRUE, frame.type = 'norm')
p1
#ggsave("clustering_by_timepoints_PCA_swab.png")
```

Clustering by treatment (PCA): after removing samples from timepoint 0 # BECAUSE WE KNOW THAT AT TIMEPOINT 0 WE SHOULD NOT HAVE DIFFERENCES BETWEEN GROUPS

```{r, echo=FALSE}
mat_swab_1 <- filter(mat_swab, timepoint != "0 before_oil")
matx0= data.matrix(select(mat_swab_1, -c(treatment,timepoint)))

p2 <- autoplot(prcomp(1-matx0), data=mat_swab_1, colour = "treatment", label = TRUE, label.size = 3, frame = TRUE, frame.type='norm')
p2
#ggsave("clustering_by_treatment_PCA_swab.png")
```

<!-- Clustering by treatment (PCA): by timepoints  -->

<!-- Time-point 0 before_oil -->

<!-- ```{r, echo=FALSE} -->
<!-- mat_swab_1 <- filter(mat_swab, timepoint =="0 before_oil" ) -->
<!-- # mat_swab_1 <- filter(mat_swab, timepoint =!"0", timepoint =!"8",  timepoint =!"21"  ) -->

<!-- matxTP= data.matrix(select(mat_swab_1, -c(treatment,timepoint))) -->

<!-- p2 <- autoplot(prcomp(1-matx0), data=mat_swab_1, colour = "treatment", label = TRUE, label.size = 3, frame = TRUE, frame.type='norm') -->
<!-- p2 -->
<!-- #ggsave("clustering_by_treatment_PCA_swab_T28.png") -->
<!-- ``` -->

<!-- Time-point 0 after_oil -->

<!-- ```{r, echo=FALSE} -->
<!-- mat_swab_1 <- filter(mat_swab, timepoint =="0 after_oil" ) -->
<!-- matxTP= data.matrix(select(mat_swab_1, -c(treatment,timepoint))) -->

<!-- p2 <- autoplot(prcomp(1-matx0), data=mat_swab_1, colour = "treatment", label = TRUE, label.size = 3, frame = TRUE, frame.type='norm') -->
<!-- p2 -->
<!-- #ggsave("clustering_by_treatment_PCA_swab_T21.png") -->
<!-- ``` -->

### MDS

#### Clustering by treatment (MDS): **NMDS1 vs NMDS2**

```{r, echo=FALSE, results='hide'}
udder.mds= metaMDS(matx0, k=3) #function metaMDS in Vegan

hull_f <- function(df) {

  temp <- data.frame(NULL)
  for (ll in unique(df$group)) {

    nn <- df[df$group == ll,][chull(df[df$group == ll, c("NMDS1","NMDS2","NMDS3")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

udder.scores <- as.data.frame(scores(udder.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
udder.scores$group <- mat_swab_1$treatment #  add the grp variable created earlier

hull.data <- hull_f(udder.scores)

g <- ggplot(data=udder.scores, aes(x=NMDS1,y=NMDS2))
g <- g + coord_equal()
g <- g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS2,fill=group,group=group),alpha=0.30)
g <- g + geom_point(data=udder.scores,aes(x=NMDS1,y=NMDS2,shape=group,colour=group),size=4)
g <- g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g <- g + ggtitle("Treatment")
g
```

#### Clustering by timepoint (MDS): 

<!-- **NMDS1 vs NMDS2** -->

<!-- ```{r, echo=FALSE, results='hide'} -->
<!-- udder.mds= metaMDS(matx, k=3) #function metaMDS in Vegan -->

<!-- hull_f <- function(df) { -->

<!--   temp <- data.frame(NULL) -->
<!--   for (ll in unique(df$group)) { -->

<!--     nn <- df[df$group == ll,][chull(df[df$group == ll, c("NMDS1","NMDS2","NMDS3")]),] -->
<!--     temp <- rbind.data.frame(temp,nn) -->
<!--   } -->
<!--   return(temp) -->
<!-- } -->

<!-- udder.scores <- as.data.frame(scores(udder.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame -->
<!-- udder.scores$group <- mat_swab$timepoint #  add the grp variable created earlier -->

<!-- hull.data <- hull_f(udder.scores) -->

<!-- g <- ggplot(data=udder.scores, aes(x=NMDS1,y=NMDS2)) -->
<!-- g <- g + coord_equal() -->
<!-- g = g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS2,fill=group,group=group),alpha=0.30) -->
<!-- g = g + geom_point(data=udder.scores,aes(x=NMDS1,y=NMDS2,shape=group,colour=group),size=4) -->
<!-- g = g + scale_fill_manual(values = c("green", "#E7B800","red","yellow")) -->
<!-- g = g + scale_color_manual(values = c("green", "#E7B800","red","yellow")) -->
<!-- g = g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) -->
<!-- g <- g + ggtitle("Source") -->
<!-- g -->
<!-- ``` -->

<!-- Clustering by type (MDS): **NMDS1 vs NMDS3** -->

<!-- ```{r, echo=FALSE, results='hide'} -->
<!-- g <- ggplot(data=udder.scores, aes(x=NMDS1,y=NMDS3)) -->
<!-- g <- g + coord_equal() -->
<!-- g = g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS3,fill=group,group=group),alpha=0.30) -->
<!-- g = g + geom_point(data=udder.scores,aes(x=NMDS1,y=NMDS3,shape=group,colour=group),size=4) -->
<!-- g = g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) -->
<!-- g <- g + ggtitle("Source ") -->
<!-- g -->
<!-- ``` -->

Clustering by type (MDS): **NMDS2 vs NMDS3**

```{r, echo=FALSE, results='hide'}
g <- ggplot(data=udder.scores, aes(x=NMDS2,y=NMDS3))
g <- g + coord_equal()
g = g + geom_polygon(data=hull.data,aes(x=NMDS2,y=NMDS3,fill=group,group=group),alpha=0.30)
g = g + geom_point(data=udder.scores,aes(x=NMDS2,y=NMDS3,shape=group,colour=group),size=4)
g = g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g <- g + ggtitle("Source")
g
```

### Clustering by treatment by timepoint (MDS):

```{r, echo=FALSE, results='hide'}
## HULL - timepoint + treatment
udder.mds= metaMDS(matx) #function metaMDS in Vegan

udder.scores <- as.data.frame(scores(udder.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
udder.scores$timepoint <- mat_swab$timepoint #  add the grp variable created earlier
udder.scores$treatment <- mat_swab$treatment


hull_f <- function(df) {
  
  temp <- data.frame(NULL)
  for (ll in unique(df$timepoint)) {
    
    nn <- df[df$type == ll,][chull(df[df$type == ll, c("NMDS1","NMDS2")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

hull.data <- hull_f(udder.scores)

g= ggplot(data=udder.scores, aes(x=NMDS1,y=NMDS2))
# g <- g + geom_point(aes(shape=group,colour=group),size=3)
# g= g + coord_equal() 
g= g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS2,fill=treatment,group=treatment),alpha=0.30) 
g= g + geom_mark_ellipse(aes(fill=treatment), con.colour = "white")
g= g + geom_point(data=udder.scores,aes(x=NMDS1,y=NMDS2,shape=treatment,colour=treatment),size=4)
g= g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g= g + facet_wrap(~timepoint, scales = "free")
# g <- g + theme(legend.position="none")
g= g + ggtitle("TREATMENT x TIMEPOINT")
g
```

## Significance of between-group distances

Significance values based on permuted analysis of variance (999 permutations), repeated 100 times.

```{r, label='permanova', echo=FALSE}
pv_treatment <- replicate(100,
  adonis2(matx ~ mat_swab$treatment, permutations = 999)$"Pr(>F)"[1],
  simplify = "vector"
)

permanova <- function(mat,term,perms) {
  
  perm <- how(nperm = sample(perms,1))
  obj <- adonis(mat ~ term, add = TRUE, parallel = TRUE, permutations = perm)
  return(obj$aov.tab$`Pr(>F)`[1])
}

pv_timepoint <- replicate(100,
  permanova(matx, mat_swab$timepoint, 999),
  simplify = "vector"
)

pv_treat_time <- replicate(100,
  adonis2(matx ~ mat_swab$treatment + mat_swab$timepoint + mat_swab$timepoint:mat_swab$treatment, permutations = 999)$"Pr(>F)"[1],
  simplify = "vector"
)

```

- between treatments: p-value = **`r mean(pv_treatment)`** (std = `r sd(pv_treatment)`)
- between types: p-value = **`r mean(pv_timepoint)`** (std = `r sd(pv_timepoint)`)
- between treatments and type =  p-value = **`r mean(pv_treat_time)`** (std = `r sd(pv_treat_time)`)


Below the ANOVA table from the model:

$$
X = \mu + treatment + timepoint + treatment*type + e
$$

```{r pairwise, echo=FALSE, message=FALSE, warning=FALSE}
source("/home/oem/milkqua_microbiome/Milk_&_SkinSwab/pairwise_adonis.r")
obj <- adonis(matx ~ mat_swab$treatment+mat_swab$timepoint)
row.names(obj$aov.tab) <- c("treatment","timepoint","residuals","total")
kable(obj$aov.tab)
```

## going 3D

```{r, echo=FALSE, warning=FALSE, message=FALSE}
## HULL - timepoint + treatment
udder.mds= metaMDS(matx, k = 3) #function metaMDS in Vegan

udder.scores <- as.data.frame(scores(udder.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
udder.scores$timepoint <- mat_swab$timepoint #  add the grp variable created earlier
udder.scores$treatment <- mat_swab$treatment


hull_f <- function(df) {
  
  temp <- data.frame(NULL)
  for (ll in unique(df$timepoint)) {
    
    nn <- df[df$timepoint == ll,][chull(df[df$timepoint == ll, c("NMDS1","NMDS2","NMDS3")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

hull.data <- hull_f(udder.scores)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
## 3D plot
p <- plot_ly(data = udder.scores, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~timepoint) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'timepoint',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p

if(pandoc_available()) {
  
  htmlwidgets::saveWidget(as_widget(p), "beta_timepoint_3d.html")
}
```

### timepoint 0 before_oil

```{r, echo=FALSE, warning=FALSE, message=FALSE}
temp <- filter(udder.scores, timepoint == "0 before_oil")

p <- plot_ly(data = temp, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

### timepoint 0 after_oil

```{r, echo=FALSE, warning=FALSE, message=FALSE}
temp <- filter(udder.scores, timepoint == "0 after_oil")

p <- plot_ly(data = temp, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```
