---
title: "Beta_diversity_Stools"
author: "Chiara Gini"
date: "12/05/2022"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
library("knitr")
library("dplyr")
library("tidyr")
library("broom")
library("vegan")
library("ggplot2")
library("reshape2")
library("rmarkdown")
library("tidyverse")
library("ggfortify")
library("data.table")
library("ggforce")
library("plotly")
library("lattice")
library("ggpubr")
```

## Beta diversity

Between-sample variability from 16S rRna-gene sequencing data.

```{r beta, echo=FALSE}
metadata <- fread ("~/Results/STOOLS/mapping_milkqua_stools.csv")
names(metadata)[1] <- "sample"
metadata$sample <- paste("sample-", metadata$MISEQ_ID, sep = "")
```

## Clustering

### PCA

```{r pressure, echo=FALSE, message=FALSE}
matrice= fread("~/Results/STOOLS/results/weighted_unifrac_distances.csv", header = T, sep = ",")
matrice  <- matrice %>% remove_rownames %>% column_to_rownames(var="row")

samples = metadata %>% pull(sample)
vec <- rownames(matrice) %in% samples
matrice = matrice[vec,vec]

matrice$treatment <- as.character(metadata$treatment[match(row.names(matrice),metadata$sample)])
matrice$timepoint <- as.character(metadata$timepoint[match(row.names(matrice),metadata$sample)])

matx= data.matrix(select(matrice, -c(treatment, timepoint)))

## MDS
mds <- cmdscale(as.dist(matx))
```

Clustering by timepoint (PCA)
```{r, echo=FALSE}
p1 <- autoplot(prcomp(1-matx), data=matrice, colour = "timepoint", label = TRUE, label.size = 3, frame = TRUE, frame.type = 'norm')
p1
ggsave("~/Results/STOOLS/results/clustering_by_timepoints_PCA_stools.png")
```

Clustering by treatment (PCA)
```{r, echo=FALSE}
p3 <- autoplot(prcomp(1-matx), data=matrice, colour = "treatment", label = TRUE, label.size = 3, frame = TRUE, frame.type='norm')
p3
ggsave("~/Results/STOOLS/results/clustering_by_treatment_PCA_stools.png")
```
Clustering by treatment (PCA): after removing samples from timepoint 0 # BECAUSE WE KNOW THAT AT TIMEPOINT 0 WE SHOULD NOT HAVE DIFFERENCES BETWEEN GROUPS

```{r, echo=FALSE}
matrice_1 <- filter(matrice, timepoint != "2")
matx0= data.matrix(select(matrice_1, -c(treatment,timepoint)))

p2 <- autoplot(prcomp(1-matx0), data=matrice_1, colour = "treatment", label = TRUE, label.size = 3, frame = TRUE, frame.type='norm')
p2
ggsave("~/Results/SKINSWABS/results/clustering_by_treatment_PCA_stools.png")
```

```{r}
figure_final <- ggarrange(p1, p2, widths=c(1, 1), labels=c("A", "B"), legend = "bottom")
print(figure_final)
ggsave("~/Results/SKINSWABS/results/clustering_merged_PCA_stools.png", plot = figure_final, width = 8, height = 4)
```

### MDS

#### Clustering by treatment (MDS): **NMDS1 vs NMDS2**

```{r, echo=FALSE, results='hide'}
stools.mds= metaMDS(matx, k=3) #function metaMDS in Vegan ##, autotransform = FALSE

hull_f <- function(df) {

  temp <- data.frame(NULL)
  for (ll in unique(df$group)) {

    nn <- df[df$group == ll,][chull(df[df$group == ll, c("NMDS1","NMDS2","NMDS3")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

stools.scores <- as.data.frame(scores(stools.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
stools.scores$group <- matrice$treatment #  add the grp variable created earlier

hull.data <- hull_f(stools.scores)

g <- ggplot(data=stools.scores, aes(x=NMDS1,y=NMDS2))
g <- g + coord_equal()
g <- g + geom_polygon(data=hull.data,aes(x=NMDS1,y=NMDS2,fill=group,group=group),alpha=0.30)
g <- g + geom_point(data=stools.scores,aes(x=NMDS1,y=NMDS2,shape=group,colour=group),size=4)
g <- g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g <- g + ggtitle("Treatment")
g
```

Clustering by type (MDS): **NMDS2 vs NMDS3**

```{r, echo=FALSE, results='hide'}
g <- ggplot(data=stools.scores, aes(x=NMDS2,y=NMDS3))
g <- g + coord_equal()
g = g + geom_polygon(data=hull.data,aes(x=NMDS2,y=NMDS3,fill=group,group=group),alpha=0.30)
g = g + geom_point(data=stools.scores,aes(x=NMDS2,y=NMDS3,shape=group,colour=group),size=4)
g = g + theme_bw() + theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))
g <- g + ggtitle("Source")
g
```

## Significance of between-group distances

Significance values based on permuted analysis of variance (999 permutations), repeated 100 times.

```{r, label='permanova', echo=FALSE}
pv_treatment <- replicate(100,
  adonis2(matx ~ matrice$treatment, permutations = 999)$"Pr(>F)"[1],
  simplify = "vector"
)

n = 10
permanova <- function(mat,term,perms) {
  
  perm <- how(nperm = sample(perms,1))
  obj <- adonis(mat ~ term, add = TRUE, parallel = TRUE, permutations = perm)
  return(obj$aov.tab$`Pr(>F)`[1])
}

pv_treatment <- replicate(n,
  permanova(matx, matrice$treatment, 999),
  simplify = "vector"
)

```

- between treatments: p-value = **`r mean(pv_treatment)`** (std = `r sd(pv_treatment)`)

Below the ANOVA table from the model:

$$
X = \mu + treatment +  + treatment*type + e
$$

```{r, echo=FALSE, message=FALSE, warning=FALSE}
source("~/milkqua_microbiome/r_scripts/support_functions/pairwise_adonis.r")
obj <- adonis(matx ~ matrice$treatment)

row.names(obj$aov.tab) <- c("treatment","residuals","total")
kable(obj$aov.tab)
```

## going 3D

```{r, echo=FALSE, warning=FALSE, message=FALSE}
stools.mds= metaMDS(matx, k = 3) #function metaMDS in Vegan

stools.scores <- as.data.frame(scores(stools.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
stools.scores$treatment <- matrice$treatment


hull_f <- function(df) {
  
  temp <- data.frame(NULL)
  for (ll in unique(df$treatment)) {
    
    nn <- df[df$treatment == ll,][chull(df[df$treatment == ll, c("NMDS1","NMDS2","NMDS3")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

hull.data <- hull_f(stools.scores)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
## 3D plot
p <- plot_ly(data = stools.scores, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = '',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p

if(pandoc_available()) {
  
  htmlwidgets::saveWidget(as_widget(p), "beta_treatment_3d.html")
}
```

###  Cows


```{r, echo=FALSE, warning=FALSE, message=FALSE}
matrice$cow <- as.character(metadata$inoculation[match(row.names(matrice),metadata$sample)])

stools.mds= metaMDS(matx, k = 3) #function metaMDS in Vegan

stools.scores <- as.data.frame(scores(stools.mds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
stools.scores$cow <- matrice$cow


hull_f <- function(df) {
  
  temp <- data.frame(NULL)
  for (ll in unique(df$cow)) {
    
    nn <- df[df$treatment == ll,][chull(df[df$cow == ll, c("NMDS1","NMDS2","NMDS3")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

hull.data <- hull_f(stools.scores)
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
p <- plot_ly(data = stools.scores, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~cow) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'cow',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

