---
title: "Taxonomy_RUMEN"
author: "Chiara Gini"
date: "21/02/2022"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
library("DT")
library("knitr")
library("tidyr")
library("broom")
library("vegan")
library("ggpubr")
library("ggplot2")
library("reshape2")
library("tidyverse")
library("data.table")
library("ggrepel")
```
 
#OVERALL COMPARISON 

```{r, echo=FALSE}
#project_folder = "~/Documents/MILKQUA"
metadata <- fread("/home/mycelium/Results/RUMEN/mapping_file_rumen.csv")
names(metadata)[1] <- "sample"

metadata$treatment[ which(metadata$treatment == "no treatment (ruminal liquid + diet)")] <- "Control"
metadata$treatment[ which(metadata$treatment == "AE1")] <- "NEO"
metadata$treatment[ which(metadata$treatment == "AE sintético 1")] <- "SEO"
metadata$treatment[ which(metadata$treatment == "gamma-terpinene")] <- "G-terpinene"
metadata$treatment[ which(metadata$treatment == "p-cymene")] <- "P-cymene"

meta_subset <- filter(metadata, treatment !="ruminal liquid")
meta_subset %>%
  group_by(treatment) %>%
  dplyr::summarise(N=n())
```

```{r, echo=FALSE}
otu <- fread("/home/mycelium/Results/RUMEN/results/taxa_summary_abs/CSS_normalized_otu_table_L2.txt", header = TRUE, skip = 1)
otu$`#OTU ID` <- gsub("^.*;","",otu$"#OTU ID")
otu <- otu %>%
  group_by(`#OTU ID`) %>%
  summarise(across(everything(), sum))

uncult <- slice(otu, 299:306, 194)
uncult$`#OTU ID` <- "Uncultured or unknown"
uncult <- uncult %>%
  group_by(`#OTU ID`) %>%
  summarise(across(everything(), sum))

otu <- otu[-c(299:306, 194), ]
otu <- rbind(otu, uncult)

otu <- gather(otu, key = "sample", value ="counts", -`#OTU ID`) %>% spread(key = `#OTU ID`, value = counts)
otu$treatment = metadata$treatment[match(otu$sample,metadata$sample)]
otu <- filter(otu, sample %in% meta_subset$sample)

## relative abundances
metadata_cols = names(meta_subset)[c(1,7)]
M <- dplyr::select(otu,-all_of(metadata_cols))
M <- M/rowSums(M)
M <- bind_cols(dplyr::select(otu, all_of(metadata_cols)),M)

M <- subset(M, treatment !="ruminal liquid")
```

```{r, echo=FALSE, fig.height=12, fig.width=12}
otu = otu %>% relocate (treatment, .after = sample)
metadata_cols = names(otu)[c(1,2)]

otu <- subset(otu, treatment !="ruminal liquid")

mO <- reshape2::melt(otu,id.vars = metadata_cols, value.name = "counts", variable.name = "genus")
mO$genus <- gsub("^.*;","",mO$genus)
mO[mO$genus == "2","genus"] <- "WPS.2"
mO$sample <- factor(mO$sample)
mO <- mO %>%
  arrange(treatment)
#mO$Cow_ID <- paste(mO$cow, mO$ref)
```

```{r, echo=FALSE, fig.height=12, fig.width=12}
m1 <- reshape2::melt(M,id.vars = metadata_cols, value.name = "counts", variable.name = "genus")
m1$genus <- gsub("^.*;","",m1$genus)
m1[m1$genus == "2","genus"] <- "WPS.2"
m1$sample <- factor(m1$sample)
m1 <- m1 %>%
  arrange(treatment)

# m1$sample <- paste(as.character(m1$sample),as.character(m1$ ),sep="-")
ordered_samples <- metadata %>%
  arrange(treatment) %>%
  dplyr::select(treatment,sample)

m1$sample <- factor(m1$sample, levels = ordered_samples$sample)

m1$treatment <- factor(m1$treatment)

m1
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
D <- m1 %>%
  group_by(genus) %>%
  summarise(N=n(), avg_abund = round(mean(counts),4), std = round(sd(counts),3)) %>%
  # filter(avg_abund>0.01) %>%
  arrange(genus,avg_abund)

genus <- D %>%
  group_by(genus) %>%
  summarize("s"=sum(avg_abund)) %>%
  arrange(desc(s)) %>%
  dplyr::select(genus,s)

D$genus <- factor(D$genus, levels = rev(genus$genus[order(genus$s)]))
D <- D %>%
  arrange(genus)

D$genus <- as.character(D$genus)

oldc <- D$genus[D$avg_abund<0.01]
newc <- rep("Lower than 1%", length(oldc))
vec <- newc[match(D$genus,oldc)]
D$genus <- ifelse(D$genus %in% oldc, "Lower than 1%", as.character(D$genus))
D$genus <- gsub("group","",D$genus)

D <- D %>%
  group_by(genus) %>%
  summarise(across(everything(), sum))

D <- D[order(-D$avg_abund),]

kable(D)
```

```{r}
q <- ggpie(D, "avg_abund", label=NULL, color = "white", fill = "genus",  legend = "right", lab.pos = "out", palette = c("red", "green", "khaki", "steelblue1","maroon","lightblue","salmon","turquoise","violet","yellow","gray","tomato","navy","pink","springgreen4","peru","sienna1","plum4","mediumblue","darkorange","brown4","gold","bisque4", "black", "blue"), font.legend=c(16, "black"), lab.font = "white", ggtheme = theme_pubr()) 
q <- q + font("xy.text", size = 16, color = "white")
q

ggsave(filename = file.path("~/Results/RUMEN/results/Figure1bis.png"), plot = q, device = "png", dpi = 250, width = 20, height = 8 )

```




```{r, echo=FALSE}
pphylatot <- ggplot(D, aes(x=factor(1), y=avg_abund, fill=genus)) + geom_bar(width=1,stat="identity")
#pphylatot <- pphylatot + facet_grid(N~treatment)
pphylatot <- pphylatot + coord_polar(theta='y', start=0) #activate this if you want a piechart
#pphylatot <- pphylatot + scale_fill_brewer(palette = "Set3") 
pphylatot <- pphylatot + geom_label_repel(aes(label = paste0(avg_abund*100, "%")), position=position_stack (vjust=0.5), size=4, show.legend = F)
pphylatot <- pphylatot + guides(fill = guide_legend(title = "Phyla")) 
pphylatot <- pphylatot + xlab("Relative abundances") + ylab("")

my_palette = get_palette(c("red", "green", "khaki", "steelblue1","maroon","lightblue","salmon","turquoise","violet","yellow","gray","tomato","navy","pink","springgreen4","peru","sienna1","plum4","mediumblue","darkorange","brown4","gold","bisque4"), length(D$genus))
pphylatot <- pphylatot + xlab("relative abundances") + ylab("percentages")
pphylatot <- pphylatot + scale_fill_manual(values = my_palette)
pphylatot <- pphylatot + theme(text = element_text(size=10),
               axis.text.x = element_text(size=8),
               # axis.text.y = element_text(size=4),
               strip.text = element_text(size = 8),
               axis.text.y = element_blank(),
               axis.ticks.y = element_blank()
               # legend.text=element_text(size=10),
               # legend.title=element_text(size=10))
)
print(pphylatot)
#ggsave(pphylatot, height = 20, width = 10, filename ="taxonomy_phylum_boxplot_rumen_tot.png")

```


#COMPARISON BETWEEN TREATMENS 

## 16S sequencing data - cow rumen samples
### Treatments 

First, let's look at phyla:

```{r, echo=FALSE, warning=FALSE, message=FALSE}
D <- m1 %>%
  group_by(phylum,treatment) %>%
  summarise(N=n(), avg_abund = round(mean(counts),4), std = round(sd(counts),3)) %>%
  filter(avg_abund>0.005) %>%
  arrange(phylum,avg_abund)

phylum <- D %>%
  group_by(phylum) %>%
  summarize("s"=sum(avg_abund)) %>%
  arrange(desc(s)) %>%
  dplyr::select(phylum,s)

D$phylum <- factor(D$phylum, levels = rev(phylum$phylum[order(phylum$s)]))
D <- D %>%
  arrange(treatment,phylum)

kable(D)
```


###boxplot abundances per treatment 

##Phyla

```{r, echo=FALSE}
D$treatment = factor(D$treatment, levels = c("Control", "NEO", "SEO", "Carvacrol", "P-cymene", "G-terpinene"), ordered = TRUE)

pphyla <- ggplot(D, aes(x=factor(1), y=avg_abund, fill=phylum)) + geom_bar(width=1,stat="identity")
pphyla <- pphyla + facet_grid(N~treatment)
#pphyla <- pphyla + coord_polar(theta='y', start=0) #activate this if you want a piechart
#pphyla <- pphyla + scale_fill_brewer(palette = "Set3") 
pphyla <- pphyla + geom_label_repel(aes(label = paste0(avg_abund*100, "%")), position=position_stack (vjust=0.5), size=4, show.legend = F)
pphyla <- pphyla + guides(fill = guide_legend(title = "Phyla")) 
pphyla <- pphyla + xlab("Relative abundances") + ylab("")
my_palette = get_palette(c("green3","gold","blue","magenta","cyan","yellow","red","gray","purple","brown","black", "darkorange1","pink1", "darkorchid1", "seagreen2"), length(unique(D$phylum)))
pphyla <- pphyla + xlab("relative abundances") + ylab("percentages")
pphyla <- pphyla + scale_fill_manual(values = my_palette)
pphyla <- pphyla + theme(text = element_text(size=10),
               axis.text.x = element_text(size=8),
               # axis.text.y = element_text(size=4),
               strip.text = element_text(size = 8),
               axis.text.y = element_blank(),
               axis.ticks.y = element_blank(),
               legend.text=element_text(size=10),
               legend.title=element_text(size=10))

print(pphyla)
#ggsave(pphyla, height = 5, width = 5, filename ="taxonomy_phylum_boxplot_rumen.png")

D %>% group_by(phylum) %>% summarise(treatment=n())

D %>% group_by(treatment) %>% summarise(phylum=n())
```


```{r}
fit = aov(avg_abund ~ treatment, data = D) ## linear model value = mu + treatment + e (y = mu + x + e)
TukeyHSD(fit, "treatment", ordered = TRUE)
```

### Treatments, all taxonomic levels

```{r, echo=FALSE,message=FALSE,warning=FALSE}
D <- D %>%
  group_by(treatment) %>%
  filter(avg_abund > 0.01) %>%
  arrange(desc(avg_abund))

phylum <- D %>%
  group_by(phylum) %>%
  summarize("s"=sum(avg_abund)) %>%
  arrange(desc(s)) %>%
  dplyr::select(phylum,s)

D$phylum <- factor(D$phylum, levels = rev(phylum$phylum[order(phylum$s)]))
D <- D %>%
  arrange(phylum)

D$taxa_level = rep("phylum", nrow(D))

otu2 <- fread("/home/mycelium/Results/RUMEN/results/taxa_summary_abs/CSS_normalized_otu_table_L2.txt", header = TRUE, skip = 1)
otu3 <- fread("/home/mycelium/Results/RUMEN/results/taxa_summary_abs/CSS_normalized_otu_table_L3.txt", header = TRUE, skip = 1)
otu4 <- fread("/home/mycelium/Results/RUMEN/results/taxa_summary_abs/CSS_normalized_otu_table_L4.txt", header = TRUE, skip = 1)
otu5 <- fread("/home/mycelium/Results/RUMEN/results/taxa_summary_abs/CSS_normalized_otu_table_L5.txt", header = TRUE, skip = 1)
otu6 <- fread("/home/mycelium/Results/RUMEN/results/taxa_summary_abs/CSS_normalized_otu_table_L6.txt", header = TRUE, skip = 1)
otu <- bind_rows(otu2,otu3,otu4,otu5,otu6)

otu <- gather(otu, key = "sample", value ="counts", -`#OTU ID`) %>% spread(key = `#OTU ID`, value = counts)

otu$treatment = metadata$treatment[match(otu$sample,metadata$sample)]
otu <- filter(otu, sample %in% meta_subset$sample)

mO <- reshape2::melt(otu,id.vars = metadata_cols, value.name = "counts", variable.name = "taxa")

mO$level <- sapply(mO$taxa, function(x) length(strsplit(as.character(x),";")[[1]]) ) 

get_taxa <- function(x,n) {
  
  z <- strsplit(x, split = ";")
  res = lapply(z, function(x) x[n])
  return(unlist(res))
}

new_taxa <- c(NULL)
for (i in seq_along(mO$counts)) {
  
  arg1 = as.character(mO$taxa[i])
  arg2 = ifelse(mO$level[i] <= 6, mO$level[i], 6)
  
  rr = get_taxa(arg1,arg2)
  new_taxa <- c(new_taxa,rr)
}

mO$new_taxa <- new_taxa

table(mO$level)
neues <- c("phylum","class","order","family","genus")
altes <- seq(2,6)
mO$level <- neues[match(mO$level,altes)]

## relative abundances
mO <- mO %>%
  group_by(level,sample) %>%
  mutate(tot = sum(counts), rel_abundance = counts/tot)

D <- mO %>%
  group_by(level,new_taxa,treatment) %>%
  summarise(avg_abund = round(mean(rel_abundance),4), std = round(sd(rel_abundance),3))
```

A table of relative abundances (and standard deviations) per taxa and treatment:

```{r, tab2, echo=FALSE}
D <- na.omit(D)
D <- D %>%
  group_by(new_taxa,treatment) %>%
  filter(avg_abund > 0.01) %>%
  arrange(level,desc(avg_abund))

kable(D)
```
A table of relative abundances in % (and standard deviations) per taxa 

```{r, echo=FALSE}
D <- na.omit(D)
D <- D %>%
  group_by(level) %>%
  filter(avg_abund > 0.01) %>%
  arrange(level,desc(avg_abund*100))

kable(D)
```

## Genera 

```{r}
Dgenera <- D %>%
  filter(level=="genus")

pgenera <- ggplot(Dgenera, aes(x=factor(1), y=avg_abund, fill=new_taxa)) + geom_bar(width=1,stat="identity", position = "fill")
# pgenera <- pgenera + facet_grid( ~treatment)
pgenera <- pgenera + guides(fill = guide_legend(title = "Genera")) 
pgenera <- pgenera + geom_label_repel(aes(label = paste0(avg_abund*100, "%")), position=position_stack (vjust=0.5), size=4, show.legend = F)
my_palette2 = get_palette(c("green3","gold","blue","magenta","cyan","yellow","red","gray","purple","brown","black", "darkorange1","pink1", "darkorchid1", "seagreen2", "darkmagenta", "burlywood1", "maroon", "darkred", "burlywood3", "chartreuse4", "coral", "peachpuff", "mediumorchid3", "yellow3", "turquoise4", "snow", "slategray3", "orangered", "olivedrab", "plum", "lightcyan", "gray38", "forestgreen", "firebrick1"), length(unique(Dgenera$new_taxa)))
pgenera <- pgenera + xlab("relative abundances") + ylab("percentages")
pgenera <- pgenera + scale_fill_manual(values = my_palette2)
pgenera <- pgenera + theme(text = element_text(size=10),
               axis.text.x = element_text(size=8),
               # axis.text.y = element_text(size=4),
               strip.text = element_text(size = 8),
               axis.text.y = element_blank(),
               axis.ticks.y = element_blank(),
               legend.text=element_text(size=5),
               legend.title=element_text(size=5))
print(pgenera)
#ggsave(pgenera, height = 8, width = 16, filename ="taxonomy_genera_bargraph_rumen.png")

Dgenera %>% group_by(new_taxa) %>% summarise(treatment=n())

Dgenera %>% group_by(treatment) %>% summarise(new_taxa=n())
```

```{r}
fit = aov(avg_abund ~ treatment, data = Dgenera) ## linear model value = mu + treatment + e (y = mu + x + e)
TukeyHSD(fit, "treatment", ordered = TRUE)
```

##Order



Same results, but in wide-format (treatments on the horizontal axis): relative abundances (no standard deviations):

```{r, echo=FALSE}
D$level  <- factor(D$level,levels = c("phylum","class","order","family","genus"))

D0 <- D %>%
  dplyr::select(-c(std)) %>%
  spread(treatment,avg_abund)

# D0[is.na(D0)] <- ""

bg.picker <- function(z) {
  
  if(is.na(z)) {return(NULL)}
  if(is.numeric(z)) {
    
    if(z > 0.05 & z <= 0.10) {return("pink")}
    if(z > 0.10) {return("yellow")}
  }
}

library("formattable")
formattable(D0, list(
  
  `Control` = color_tile("white","green"),
  `Treated` = color_tile("white","deepskyblue")
  # `3 PR` = color_tile("white","yellow"),
  # `4 PR` = color_tile("white","pink"),
  # `5 PR` = color_tile("white","yellow"),
  # `6 PR` = formatter("span",
  #                 style = x ~ style(
  #                   "border-radius" = "4px",
  #                   "padding-right" = "4px",
  #                   "background-color" = sapply(x, bg.picker))
  #                 )
))

```

Finally,a bubble chart to visualize relative abundances per taxa and treatment:

```{r, echo=FALSE, fig.height = 14, fig.width = 10}
taxa <- D %>%
  group_by(level,new_taxa) %>%
  summarize("s"=sum(avg_abund)) %>%
  arrange(desc(s)) %>% #originally was arrange(desc(level, s)) but stopped working reporting that desc() must be called with exactly one argument
  dplyr::select(level,new_taxa,s)

taxa <- taxa %>%
  group_by(level) %>%
  arrange(level,s)

p <- ggplot(D, aes(x = treatment, y = new_taxa))
p <- p + geom_point(aes(size = avg_abund, colour = treatment), alpha = 0.4)
p <- p + facet_grid(level~treatment, scales = "free", space = "free")
p <- p + scale_color_manual(values = c("green3","gold","blue","magenta","cyan","yellow", "green", "coral",  "#E7B800", "#FC4E07"))
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1))
p <- p + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
p <- p + theme(axis.text.x = element_text(size=8))
p <- p + theme_bw()
p <- p + xlab("treatments") + ylab("average abundances")
p
#ggsave (p, height = 13, width = 8, filename = "taxonomy_bubbleplot_rumen.png")
```

### Significance of treatments
 
The significance of differences between treatments (Control vs Treated with Essential Oil) was evaluated using a linear model (ANOVA) that included the effect of  s:
 
$$
counts_{ikj} = \mu +  _k + treatment_j + e_{ikj}
$$
# 
This way, the variability due to region is removed when assessing the effect of treatments.
 
```{r, echo=FALSE}
m1 <- mO %>%
  filter(!is.na(level)) %>%
  arrange(level,treatment)


D <- m1 %>%
  group_by(level, new_taxa) %>%
  do(tidy(anova(lm(counts ~  treatment, data = .)))) %>%
  filter(term == "treatment")

D$level  <- factor(D$level,levels = c("phylum","class","order","family","genus"))
D <- D %>%
  arrange(level,new_taxa)

datatable(D, options = list(pageLength=100)) %>%
  formatStyle('p.value', backgroundColor = styleInterval(0.05, c('yellow', 'white')))

write.csv(D, "one.csv")
```


```{r, echo=FALSE}
DX <- D %>%
  filter(`p.value` <= 0.05) %>%
  dplyr::select(c(level,new_taxa, `p.value`)) %>%
  arrange(level,`p.value`)

D0 <- mO %>%
  dplyr::group_by(level,new_taxa, treatment) %>%
  dplyr::summarise(avg_counts = mean(counts))

to_save = list(D,DX,D0)
save(to_save, file = "taxonomy_ .RData")

load("taxonomy_ .RData")
D <- to_save[[1]]
DX <- to_save[[2]]
D0 <- to_save[[3]]

dd <- spread(D0, key = treatment, value = avg_counts)
temp <- inner_join(DX,dd, by = c("level" = "level", "new_taxa" = "new_taxa"))
fwrite(temp, file = "~/Results/RUMEN/results/rumen_significant_otus.csv", col.names = TRUE, sep = ",")
print (dd)
```

```{r, echo=FALSE, warning=FALSE}
load("taxonomy_ .RData")
D <- to_save[[1]]
DX <- to_save[[2]]
D0 <- to_save[[3]]

D0 <- mutate(D0, avg_counts = avg_counts+1) %>% spread(key = treatment, value = avg_counts)

D1 <- DX %>%
  inner_join(D0, by = c("level" = "level", "new_taxa" = "new_taxa")) %>%
  mutate(p.value = -log10(p.value)) %>%
  gather(key = "treatment", value = "counts", -c(level,new_taxa, p.value))

D1$level <- factor(D1$level, levels = c("phylum","class","order","family","genus"))

D1 <- D1 %>% group_by(level) %>% mutate(tot = sum(counts), relab = counts/tot)

D1 <- filter(D1, level == "genus")

D1$treatment <- factor(D1$treatment, levels = c("Control","NEO","SEO","Carvacrol","G-terpinene","P-cymene"))

p <- ggplot(D1, aes(x = treatment, y = new_taxa))
p <- p + geom_tile(aes(fill = relab), colour = "white")
p <- p + facet_grid(level~treatment, scales = "free") + xlab("Treatments") 
p <- p + scale_fill_gradient(low = "orange", high = "seagreen","pvalue", limits=c(0, 0.05))
p <- p + theme(strip.text.y = element_text(size = 11), 
               strip.text.x = element_text(size = 11),
               # axis.text.y = element_text(size = 4),
               axis.text.x = element_text(size = 11),
               axis.title = element_text(size = 11))
p <- p  + theme(axis.title.y = element_blank(),
                                    axis.text.y = element_blank(),
                                    axis.ticks.y = element_blank())
p

dd <- filter(D1, treatment == "NEO") %>% mutate(variable = "p-value")

dd <- filter(dd, level == "genus")

q <- ggplot(dd, aes(x = factor(1), y = new_taxa, group=level))
q <- q + geom_tile(aes(fill = p.value), colour = "white")
q <- q + facet_grid(level~variable, space="free", scales = "free_y") + ylab("Genera") 
q <- q + scale_fill_gradient(low = "orange", high = "seagreen")
q <- q + theme(strip.text = element_text(size = 11), 
               strip.text.x = element_text(size = 11),
               axis.text.y = element_text(size = 11),
               axis.title = element_text(size = 11))
q <- q + guides(fill=FALSE) + theme(
  # axis.title.x = element_blank(),
  # axis.text.x=element_blank(),
  # axis.ticks.x=element_blank(),
  strip.text.y = element_blank(),
  # axis.text.x = element_blank()
  axis.text.x = element_text(size = 11)
)
q <- q + xlab("") 
q

# figure_final <- ggarrange(NULL, q, NULL, p, widths=c(0.05, 0.5, 0.05, 1), labels=c("","A","", "B"), label.x = -0.05, common.legend = TRUE, legend = "bottom", ncol = 4, nrow = 1)

figure_final <- ggarrange(q, p, widths=c(0.8, 1.2), labels=c("A","B"), label.x = -0.015, common.legend = TRUE, legend = "right")
figure_final
# print(figure_final)
ggsave(filename = "~/Results/RUMEN/results/heatmap_rumen.png", plot = figure_final, device = "png", width = 12, height = 3)
```

### Extracting estimated coefficients for "significant" OTUs

```{r}
res = data.frame("level"=NULL, "OTU"=NULL, "p_value"=NULL, "Carvacrol"=NULL,"EO1"=NULL,"p-cymene"=NULL,"Synthetic EO1"=NULL,"γ-terpinene"=NULL)

for (name in DX$new_taxa) {
  
  print(paste("analysing OTU ", name))
  pval = as.numeric(DX[DX$new_taxa==name,"p.value"])
  level = DX[DX$new_taxa==name,"level"]
  
  ## estimating coefficients
  temp = filter(m1, new_taxa == name)
  temp$treatment <- factor(temp$treatment, levels = c("Control","Carvacrol","EO1","p-cymene","Synthetic EO1","γ-terpinene"))
  g = lm(counts ~ treatment, data = temp)
  
  ## extracting coefficients
  coefs = g$coefficients
  coefs = coefs[!grepl("(Intercept)", names(coefs))]
  names(coefs) = gsub("treatment","",names(coefs))
  coefs = as.data.frame(t(coefs))
  
  ## adding metadata
  coefs["level"] = level
  coefs["OTU"] = name
  coefs["p_value"] = pval

  # saving results
  res = rbind.data.frame(res, coefs)
}
```



#Check the pvalue of everytreatment against control, genus by genus

ADD A FOR LOOP HERE ASAP

```{r}
library("broom")
library("tidyverse")

load("~/Results/RUMEN/m1.Rdata")
m1 = to_save[[1]]
unique(m1$treatment)
m1$treatment = factor(m1$treatment, levels = c("Control","Carvacrol","G-terpinene", "NEO", "P-cymene","SEO"))

temp = filter(m1, level == "genus", new_taxa == "Lachnospiraceae UCG-010") ##specify genus here
fit <- lm(counts ~ treatment, data = temp)
anova(fit)
tt = summary(fit)
names(tt)

tt = tidy(fit)
filter(tt, term != "(Intercept)") %>% pull(p.value)
```

```{r}
load("~/Results/RUMEN/m1.Rdata")
m1 = to_save[[1]]
unique(m1$treatment)
m1$treatment = factor(m1$treatment, levels = c("Control","Carvacrol","G-terpinene", "NEO", "P-cymene","SEO"))

temp = filter(m1, level == "genus", new_taxa == "Synergistes") ##specify genus here


temp1= filter(temp, treatment == "SEO", level== "genus")
mean(temp1$counts)
mean(temp1$rel_abundance)
```

