---
title: "Alpha_diversity_swab"
author: "Chiara Gini"
date: "December 10, 2021"
output:
  html_document:
    df_print: paged
  ioslides_presentation: default
  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("DT")
library("knitr")
library("dplyr")
library("tidyr")
library("broom")
library("vegan")
library("scales")
library("ggplot2")
library("ggpubr")
# library("madbito")
library("reshape2")
library("gghighlight")
library("data.table")
```

## Alpha diversity of milk samples

From 16S rRNA-gene sequencing of cow milk microbiome samples:

- two treatments for skin samples: before and after oil application

Main alpha diversity indexes per group. 

```{r alpha, echo=FALSE}
metadata <- fread("/home/mycelium/milkqua_project/Mapping_file/mapping_file_def.csv")
names(metadata)[1] <- "sample"

alpha <- read.table("/home/mycelium/milkqua_project/qiime_1.9_results/results/alpha_diversity/alpha.txt", header = TRUE)
alpha$sample <- row.names(alpha)
alpha$observed_species <- NULL

mAlpha <- reshape2::melt(alpha, id.vars = "sample", variable.name = "metric", value.name = "value")

mAlpha$type <- metadata$sample_type[match(mAlpha$sample,metadata$sample)]
mAlpha$timepoint <- metadata$timepoint[match(mAlpha$sample,metadata$sample)]
mAlpha$treatment <- metadata$treatment[match(mAlpha$sample,metadata$sample)]
mAlpha <- filter(mAlpha, type == "swab") %>% select(-c(type))
```

## Averages and plots

```{r pressure, echo=FALSE}
  mAlpha$timepoint <- factor(mAlpha$timepoint, levels = c("before_oil","after_oil"))
D <- mAlpha %>%
  group_by(metric,treatment, timepoint) %>%
  summarize(N=n(),avg=round(mean(value),3)) %>%
  spread(key = metric, value = avg)

D$timepoint <- factor(D$timepoint, levels = c("before_oil","after_oil"))

d_alpha <- D %>%
  gather(key = "metric", value = "value", -c(treatment,timepoint,N)) %>%
  spread(key = "treatment", value = "value")

kable(D)
```

### Boxplots of alpha diversity indexes

#### Treatment by region

```{r, echo=FALSE, fig.height=12, fig.width=12}
mAlpha <- mAlpha %>% group_by(metric) %>%
  mutate(scaled_value = rescale(value, to = c(0,100)))
  
p <- ggplot(mAlpha, aes(x=timepoint,y=scaled_value)) 
p <- p + geom_boxplot(aes(fill=treatment))
p <- p + scale_fill_manual(values=c("#56B4E9", "#E69F00"))
p <- p + facet_wrap(~metric, scales = "free")
p <- p + xlab("sample type")
p = p + theme(strip.text = element_text(size = 12),
              axis.text.x = element_text(size = 8))
print(p)  
ggsave ("boxplot_alpha_diversity_treatmentby timepoint_swab.png")
```




## Testing for significance

### Original alpha-diversity index-values - effect of treatment in swab samples

```{r test_alpha, echo=FALSE}
# within(mAlpha, treatment <- relevel(treatment, ref = "carn")

D <- mAlpha %>%
  group_by(metric, timepoint) %>%
  do(tidy(lm(value ~ treatment , data = .))) %>%
  filter(term!= "(Intercept)")

datatable(D, options = list(pageLength=100)) %>% 
  formatStyle('p.value', backgroundColor = styleInterval(0.05, c('yellow', 'white')))

```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- ggplot(D, aes(x=timepoint, y=p.value))
p <- p + geom_jitter(aes(group=metric, colour=metric), width = 0.2)
p <- p + gghighlight(p.value < 0.10)
p <- p + xlab("timepoint")
p

ggsave(filename = "alpha_significance_swab.png", plot = p, device = "png")
```

```{r}
# inner_join(x = d_alpha, y = select(D, c(type, metric, p.value)), by = c("type" = "type","metric" = "metric")) %>% fwrite(file = "alpha_table.csv", sep=",")
```

```{r}
## Correction for baseline
bl_medie <- mAlpha %>%
  filter(timepoint=="before_oil") %>%
  group_by(metric,treatment) %>%
  summarize(media_bl=mean(value))
  
bl_counts <- mAlpha %>%
  group_by(metric,treatment) %>%
  filter(timepoint=="before_oil") %>%
  arrange(metric) %>%
  rename(value.bl = value)

M <- merge(mAlpha,bl_counts[,c(1,2,3)],by=c("metric","sample"),all.x = TRUE)
M1 <- M%>%
  filter(!is.na(M$value.bl))


M2 <- M %>%
  filter(is.na(value.bl)) %>%
  mutate(value.bl=replace(value.bl,is.na(value.bl),right_join(bl_medie, ., by =c("metric","treatment"))$media_bl))

M <- rbind.data.frame(M1,M2)  

M <- M %>%
  mutate(corrected_counts=value-value.bl) %>%
  filter(value!=0) %>%
  arrange(sample)

M <- M %>%
  group_by(metric) %>%
  mutate(scaled_counts=scales::rescale(corrected_counts,c(0,100)))
```

```{r, echo=FALSE}
D <- M %>%
  group_by(metric,treatment,timepoint) %>%
  summarize(avg=mean(corrected_counts)) %>%
  spread(key = metric, value = avg)

kable(D)
```


```{r, echo=FALSE}
mD <- melt(D, id.vars = c("timepoint","treatment"), variable.name = "metric")

p <- ggplot(mD, aes(x=timepoint,y=value, group=treatment)) + geom_point(aes(colour=treatment))
p <- p + geom_line(aes(colour=treatment))
p <- p + facet_wrap(~metric, scales = "free_y")
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1))
p  

# ggsave(filename = "alpha_significance_by_timepoints_swab.png", plot = p, device = "png")
```

```{r test_alpha_bl, echo=FALSE}
# within(mAlpha, treatment <- relevel(treatment, ref = "carn"))

D <- M %>%
  filter(timepoint != 0) %>%
  group_by(metric, timepoint) %>%
  do(tidy(lm(value ~ treatment, data = .))) %>%
  filter(term != "(Intercept)")

datatable(D, options = list(pageLength=100)) %>% 
  formatStyle('p.value', backgroundColor = styleInterval(0.05, c('yellow', 'white')))

```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- ggplot(D, aes(x=timepoint, y=p.value))
p <- p + geom_jitter(aes(group=metric, colour=metric), width = 0.2)
p <- p + gghighlight(p.value < 0.10)
p <- p + xlab("timepoint")
p

ggsave(filename = "alpha_significance_skinswab.png", plot = p, device = "png")
```









