---
title: "alpha_diversity"
author: "Filippo Biscarini"
date: "November 08, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("DT")
library("knitr")
library("dplyr")
library("tidyr")
library("broom")
library("vegan")
library("scales")
library("ggplot2")
library("ggpubr")
# library("madbito")
library("reshape2")
library("gghighlight")
library("data.table")
```

## Alpha diversity

From 16S rRNA-gene sequencing of cow milk and udder skin microbiome samples:

- two treatments for skin samples: before and after oil application
- 4 timepoints for milk samples and two treatments (oil, treated/untreated)

Main alpha diversity indexes per group. 

```{r alpha, echo=FALSE}
metadata <- fread("mapping_file.csv")
names(metadata)[1] <- "sample"

alpha <- read.table("results/alpha_diversity/alpha.txt", header = TRUE)
alpha$sample <- row.names(alpha)
alpha$observed_species <- NULL

mAlpha <- reshape2::melt(alpha, id.vars = "sample", variable.name = "metric", value.name = "value")

mAlpha$type <- metadata$sample_type[match(mAlpha$sample,metadata$sample)]
mAlpha$timepoint <- metadata$timepoint[match(mAlpha$sample,metadata$sample)]
mAlpha$treatment <- metadata$treatment[match(mAlpha$sample,metadata$sample)]
mAlpha <- filter(mAlpha, type == "milk") %>% select(-c(type,treatment))
```

## Averages and plots

```{r pressure, echo=FALSE}
D <- mAlpha %>%
  group_by(metric,treatment,type, timepoint) %>%
  summarize(N=n(),avg=round(mean(value),3)) %>%
  spread(key = metric, value = avg)

d_alpha <- D %>%
  gather(key = "metric", value = "value", -c(treatment,type,timepoint,N)) %>%
  spread(key = "treatment", value = "value")

kable(D)
```

### Boxplots of alpha diversity indexes

#### Treatment by region

```{r, echo=FALSE, fig.height=12, fig.width=6}
mAlpha <- mAlpha %>% group_by(metric) %>%
  mutate(scaled_value = rescale(value, to = c(0,100)))
  
p <- ggplot(mAlpha, aes(x=type,y=scaled_value)) 
p <- p + geom_boxplot(aes(fill=type))
p <- p + scale_fill_manual(values=c("#56B4E9", "#E69F00"))
p <- p + facet_grid(metric~timepoint, scales = "free", space = "free")
p <- p + xlab("sampke type")
print(p)  
```




## Testing for significance

### Original alpha-diversity index-values - effect of timepoint in milk samples

```{r test_alpha, echo=FALSE}
milk <- filter(mAlpha, type == "milk")
# within(mAlpha, treatment <- relevel(treatment, ref = "carn"))

D <- milk %>%
  group_by(metric) %>%
  do(tidy(lm(value ~ timepoint, data = .))) %>%
  filter(term != "(Intercept)")

datatable(D, options = list(pageLength=100)) %>% 
  formatStyle('p.value', backgroundColor = styleInterval(0.05, c('yellow', 'white')))

```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- ggplot(D, aes(x=term, y=p.value))
p <- p + geom_jitter(aes(group=metric, colour=metric), width = 0.2)
p <- p + gghighlight(p.value < 0.10)
p <- p + xlab("timepoint")
p

# ggsave(filename = "alpha_significance.png", plot = p, device = "png")
```

```{r}
# inner_join(x = d_alpha, y = select(D, c(type, metric, p.value)), by = c("type" = "type","metric" = "metric")) %>% fwrite(file = "alpha_table.csv", sep=",")
```




